### ISSUE TITLE
Implement dice roller for RichEditor and Game Message

### ISSUE THREAD

**Issue by mikelax** [mikelax/forgingadventures#75]
Overview
===
One of the required features in order to allow users to play games is a dice roller. This will allow the user to create one or more rolls using commonly accepted dice syntax. 

Examples
===
[Here is the BGG dice roller](https://boardgamegeek.com/wiki/page/Geek_Randomizer_Guide) guide. You can see it supports a lot of different formats. Some of the commons one are: 

* 2d6+3 - roll 2 6 side and add 3
* 2d20l+3 - roll 2 twenty side and drop the lowest and then add 3
* 1custom3{Tom;Dick;Harry} - roll 1 custom 3 side, using the values Tom, Dick, Harry
* Exploding dice
* Fudge dice

Existing Packages
===
* [chance](https://www.npmjs.com/package/chance) - This is a mature package that focuses in trying to get close to real random. The `rpg` and `dice` rolls are only a really small part of the overall package. Probably not focused enough on just dice rolls. 
* https://www.npmjs.com/package/d20
* https://www.npmjs.com/package/roll

Out of the above libs, I think the [roll](https://www.npmjs.com/package/roll) package is closest to what we are looking for. 

Requirements
===
* A User should be able to add multiple dice rolls to a single message. 
* The User should be able to add a descriptive label to each roll.
* Once a message is saved, the dice rolls should be able to be edited (ie. they are immutable).
  * A nice to have would be to not allow the user to remove any dice rolls once saved, though I don't know if this is required for MVP. 
* Add a new column to game_messages table to store an array of dice rolls. Each dice roll object will be an object of:  label, result, components of roll
* We do only need to implement some of the basic dice modifiers as this feature certainly follows the 80/20 rule. 20% of the dice features will be used 80% of the time. 
These features are:  
* multiple dice, add/subtract constants
* drop high / drop low
* custom dice

Future Considerations
===
Once we have some basic character attributes we can use those values to automatically calculate certain rolls. These rolls will have to be based on the game label. For example for d&d5e, to roll `Initiative` is 1d20 plus the character `Dexterity` modifier. So the user could say:  `/roll Initiative`

We will only be able to easily calculate certain rolls. We could also allow the user to create their own roll templates, or bookmarks. For example:  `/roll 2d20l+3 save advantage-attack`
they could then later use:  `/roll advantage-attack`

**Comment by mikelax** [comment]
@nazar  as a general question, do you think that we should implement this as a graphql mutation or create a REST endpoint to create a roll? This question might apply to some other functions that aren't totally graphql related and might lend themselves more to a REST Post.

**Comment by nazar** [comment]
>do you think that we should implement this as a graphql mutation or create a REST endpoint to create a roll? 

I think graphQL all the way unless there are technical limitations the prohibit us using it - i.e. lack of file graphQL types and the need for a multer for image uploads.

Do keep in mind that we can use the dice rolling library in both the client (to validate and preview roll dice results for example) and in the API to run the actual dice roll when the message is submitted.

I was thinking we'd do this in a modal that is triggered from the editor action toolbar where the user can either type out the dice rolls and get a preview (i.e. 2d6 +2       - sample result: 9) or build a simple interface with drops downs (i.e. no of dice dropdown, dice sides dropdown, bonus numeric input).

Once the user hits Save on the modal, we construct a meta object and save it to `game_messages.meta` with a JSON structure with something like:

```
meta = {
  rolls: [
    {
      input: "3d20 + 5",
      label: "attack roll"
      result: calculated on the api and saved here
    }, {
      input: "1d20 - 5",
      label: "damage vs orc",
      result: calculated on the api and saved here
    }
  ]
}

```

**Comment by mikelax** [comment]
@nazar  that all sounds pretty good. 
for a reference I think the modal idea is good, we would just need to be "at least" as good as this UI. 
We would need to support a "label", or description for the roll, especially if they will be displayed as meta attachments. 
I don't think that we should bother trying to create all the dropdowns, there will just be too many combinations to make that worth it. 

<img width="593" alt="screen shot 2018-02-21 at 10 26 44 am" src="https://user-images.githubusercontent.com/2080701/36488676-0dc062fa-16f2-11e8-8989-09f9cd837cb6.png">

This site allows you to preview a post, in our case this would just be displayed in the editor as the user is creating/editing the message. Basically we could display the "syntax" entered for the roll, but only display the result after the message is created. 
<img width="664" alt="screen shot 2018-02-21 at 10 34 01 am" src="https://user-images.githubusercontent.com/2080701/36488966-cb395760-16f2-11e8-9259-896a376d4db3.png">

**Comment by mikelax** [comment]
@nazar  after thinking about this a bit more, I think your idea of the meta attachments for the rolls is a good idea. In looking at this more it seems that if we go the route of using the meta column, the work is more related to the editor than not. It makes sense to get the image uploads sorted and finalized before starting this task. 

Also, we should try to isolate the "dice roll" logic the best we can, so if we do decide to switch libs, etc. code updates are minimal.
